include/device.h

```diff
@@ -220,6 +260,8 @@ struct device {
 	const void *api;
 	/** Address of the device instance private data */
 	void * const data;
+	/** optional pointer to handles associated with the device */
+	const device_handle_t *const handles;
```

Comment:
Can't we get a more self-describing attribute name? (like childs, which is what these are no?)

And make the comment better describing the association. How is this associated? Or of course if the attribute is self describing, no need to tell more.

Reply:
The array contains a sequence of sets of nodes.  Not all are child nodes; in the initial pass the first node is the self identifier and the reset are nodes it depends on (whether child or arbitrary, such as a chip-select GPIO).  If we need to know the node's parent device, it'd go here too.  I don't have a better name than "handles".

I've updated the description to make this more clear.

Reply:
> The array contains a sequence of sets of nodes 

It's a simple pointer, so by itself it's a sequence of nodes. (sure, each node, translated to an actual device can point to another sequence etc... each device's handles attribute is a sub-tree then). 

> Not all are child nodes; in the initial pass the first node is the self identifier and the reset are nodes it depends on

So none are child nodes, since it cannot depend on its own child? Seems like it points to parent and/or dependency nodes only then. 
What do you mean by "the initial pass"? is that related to the state of the array before the elf script?

> If we need to know the node's parent device

How do you differentiate between a dependency node and its actual parent?

Maybe I missed it, but do you have a visual (ascii art, or picture etc..) of what the tree could look like? I think it will explain better.

Reply:
> What do you mean by "the initial pass"? is that related to the state of the array before the elf script?

Yes.

> > The array contains a sequence of sets of nodes
> 
> It's a simple pointer, so by itself it's a sequence of nodes. (sure, each node, translated to an actual device can point to another sequence etc... each device's handles attribute is a sub-tree then).

No, it really is (a pointer to) a sequence of sets of nodes, where each set is a sequence of handles and the sets are separated by a special value, and the whole thing terminated by another special value.  I expect the specific sets encoded in the sequence to change over time as we need more or less information from devicetree.

> > Not all are child nodes; in the initial pass the first node is the self identifier and the reset are nodes it depends on
> 
> So none are child nodes, since it cannot depend on its own child? Seems like it points to parent and/or dependency nodes only then.

For now, yes.  In general, it's whatever device handles are relevant and definable at build time.

> > If we need to know the node's parent device
> 
> How do you differentiate between a dependency node and its actual parent?

Right now, we don't.  Nothing needs the parent as a distinct identifiable entity.

The dependencies are from the devicetree analysis, exposed by [this macro](https://docs.zephyrproject.org/latest/reference/devicetree/api.html?highlight=ordinals#c.DT_SUPPORTS_DEP_ORDS).  Basically it's the parent node and any node that is directly referenced by a property in the node.  That may change as we get more experience with what dependencies are needed, but however it's done they're all device references that will end up in this sequence.

> Maybe I missed it, but do you have a visual (ascii art, or picture etc..) of what the tree could look like? I think it will explain better.

No, I don't.  There's one in @nashif's power management architecture presentation, which I can't find right now and the link has expired from slack.  But the sorts of things that aren't parents include:
* gpio phandles, e.g. to the device that controls the signal for chip-select, reset, interrupt, ...
* parent bus devices
* clock or power domain devices, such as PWM timers and voltage/current regulators

The dependency relations define a directed acyclic graph, not a tree.

Reply:
I had made this:
```
         * This encodes a sequence of sets of device handles that have
￼	 * some relationship to this node.  The individual sets are
￼	 * extracted with dedicated API, such as
￼	 * device_get_requires_handles().
```

I can't say what the sets will be because I don't know yet: this is a core capability for which applications will be identified over time.  But the values in this sequence are handles or non-handle marker values that delineate the ordered sequence of handle sets.

Could you suggest wording that describes this another way?